stl/inc/bit

```diff
@@ -151,16 +139,57 @@ _NODISCARD int _Checked_x86_x64_countl_zero(const _Ty _Val) noexcept {
         const unsigned int _High = _Val >> 32;
         const auto _Low          = static_cast<unsigned int>(_Val);
         if (_High == 0) {
-            return 32 + _Checked_x86_x64_countl_zero(_Low);
+            return 32 + _x86_x64_countl_zero_lzcnt(_Low);
         } else {
-            return _Checked_x86_x64_countl_zero(_High);
+            return _x86_x64_countl_zero_lzcnt(_High);
         }
 #else // ^^^ _M_IX86 / !_M_IX86 vvv
         return static_cast<int>(__lzcnt64(_Val));
 #endif // _M_IX86
     }
-    // note: we don't need to call a fallback here because
-    // all supported x86 processors at least have bsr/bsf
+}
+
+template <class _Ty>
+_NODISCARD int _x86_x64_countl_zero_bsr(const _Ty _Val) noexcept {
+    constexpr int _Digits = numeric_limits<_Ty>::digits;
+
+    unsigned long _Result;
+    if constexpr (_Digits <= 32) {
+        if (!_BitScanReverse(&_Result, _Val)) {
+            return _Digits;
+        }
+    } else {
+#ifdef _M_IX86
+        const unsigned int _High = _Val >> 32;
+        if (_BitScanReverse(&_Result, _High)) {
+            return static_cast<int>(31 - _Result);
+        }
+
+        const auto _Low = static_cast<unsigned int>(_Val);
+        if (!_BitScanReverse(&_Result, _Low)) {
+            return _Digits;
+        }
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        if (!_BitScanReverse64(&_Result, _Val)) {
+            return _Digits;
+        }
+#endif // _M_IX86
+    }
+    return static_cast<int>(_Digits - 1 - _Result);
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_countl_zero(const _Ty _Val) noexcept {
+#ifdef __AVX2__
+    return _x86_x64_countl_zero_lzcnt(_Val);
+#else
+    const bool _Have_lzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    if (_Have_lzcnt) {
+        return _x86_x64_countl_zero_lzcnt(_Val);
+    } else {
+        return _x86_x64_countl_zero_bsr(_Val);
+    }

```

Comment:
I think this logic is robust and simple to understand, thanks! Based on what we've learned about CPU history, my last concern is whether `_Have_lzcnt`'s name (which was pre-existing) is now confusing (because `true` is "definitely have" and `false` is "not sure, must fallback"). It looks like this also applies to `_Have_popcnt` (as some processors supported it before SSE 4.2?) and it definitely applies to `_Have_tzcnt`.

Should we take this opportunity to rename each bool from `_Has_meow` to `_Definitely_has_meow`, which I believe would indicate their meaning? I would also be fine with leaving this as-is since it seems unlikely we'll forget soon (and by the time we do, the relevant processors will all have crumbled into sand).

Reply:
Done

Reply:
`popcnt` situation, according to my wikipedia reading:
 * SSE4.2 is a complete version of SSE4, it includes `popcnt`
 * AMD has introduced `popcnt` in `SSE4a` which is more than SSE4.1, but less than SSE4.2

So, if I got it right AMD 10h processors are example of processors with `popcnt` but without full SSE4.2

Reply:
https://en.wikipedia.org/wiki/SSE4#POPCNT_and_LZCNT